### ISSUE TITLE
Projen Vision for Monorepos (workspaces / lerna / custom solutions)

### ISSUE THREAD

**Issue by andrestone** [projen/projen#450]
Hi there!

It's been six months since this topic first came up. I was starting a new project back then and ended up using a fork based on [this PR](https://github.com/projen/projen/pull/13), which was enough for what I needed and fairly simple to maintain since Projen was just a newborn at the time.

Now, Projen has evolved a lot (I've been trying to digest the last 6 months of merged PR's) and I can see concepts like project components and subprojects stabilizing. I want to look at these concepts as lower level abstractions that could be leveraged to build Projen project types that support popular setups to manage monorepos, such as Lerna + yarn workspaces. On the other hand, some folks don't like Lerna and using yarn workspaces might be overkill / misfit for their projects.

So the question is: do we want higher level abstractions / project types with sane defaults for managing monorepos? What would those look like?

**Comment by eladb** [comment]
Yes, we totally want to support monorepos and some of the work we've done in the past weeks (like abstractions for dependencies) is related. I am planning to pick this up soon. Stay tuned.

**Comment by skyrpex** [comment]
Is this a good place to talk about different proposals/interfaces/options?

**Comment by skyrpex** [comment]
I've been having fun last night with "curiously recurring" projen monorepo. Pretty happy with what can be done with it.

Roughly, this is what I did:
- Created a normal lerna monorepo and added projen
- Then, used esno to run  projen code using ts files
- After that, started composing projen components in the monorepo itself
- Finally, made all packages managed the projenrc file

Had to do some interesting things:
- I wanted to use ts files for the projen components, so the root project uses `projenCommand: "yarn esno $(yarn bin projen)"`. It works if I run projen as `yarn projen`
- All subprojects were trying to run `yarn install` and was pretty slow. Only the root project needs to do that. To fix that, I used a custom NodeProject with this hack: ```Object.assign(this.package, {
      renderInstallCommand: () => "echo 'Skip installing dependencies'",
    });```
- When releasing a new version, I wanted to stick with `lerna version`. Had to add a `version` npm script (that runs after lerna bumps but before commit) that copied all package.json versions to the related version.json file, runs `projen`, and then commits everything

**Comment by Chriscbr** [comment]
Commenting to keep fresh

**Comment by Chriscbr** [comment]
Keep

**Comment by moltar** [comment]
Here's my attempt at creating a monorepo project type: https://github.com/moltar/projen-turborepo

It uses [Turborepo](https://turborepo.org/docs/getting-started), which IMO will take the community by storm shortly, because:

1. It is MEGA fast (uses Go)
2. It was acquired by Vercel, which is good at hyping up the projects

A big missing piece is inability to add GitHub Actions from sub-projects. Issue here, if anyone wants to discuss: https://github.com/moltar/projen-turborepo/issues/2

**Comment by brentryan** [comment]
Any updates on plans to support monorepos? 

pnpm workspace?
nx.dev?
lerna?
turborepo?

It seems that the current project api makes some assumptions that doesn't work well for some monorepo structures so curious what the plans/direction will be.

**Comment by damnhandy** [comment]
Is the general consensus here that a monorepo in this context is always a collection of NodeJS based projects, or a mix of project types? For example, I could have a Lamba in Go, container in based on Java, and infrastructure in Typescript all in a single repo. Or is such a thing out of scope for this feature?

**Comment by mrgrain** [comment]
There is a lot of scope for what a monorepo is. I'm wondering how much projen should concern itself with this. 🤔 

Saying that, we can already see the community coming up with implementations. E.g.
https://github.com/aws/aws-prototyping-sdk/tree/mainline/packages/nx-monorepo
What do people think about this one for example?

**Comment by moltar** [comment]
I am working on a prototype for a generic "monorepo tools" package. The scope is to combine monorepo-related components in one place.

The idea is to apply components to a root package, adding functionality as desired.

This will replace too narrowly scoped https://github.com/moltar/projen-turborepo

Need a few more iterations to release it.

**Comment by agdimech** [comment]
FYI: I have implemented a monorepo using NX: https://aws.github.io/aws-prototyping-sdk/typescript/nx-monorepo/index.html

There is a sample project which can be found here: https://github.com/aws/aws-prototyping-sdk/tree/prototyping_show

**Comment by mrgrain** [comment]
Related: https://github.com/projen/projen/issues/1433

**Comment by mrgrain** [comment]
I've been playing around with a yarn monorepo recently and it's working quite well. I've found a few bits that don't work quite well yet with subprojects in general, but also with Node.js projects in general. Will try to publish this at some point.

You've might already seen some fixes coming in, expect more!

Generally speaking, I think the answer will mostly likely be a mix of making subprojects more workable, as well as specific monorepo implementations  the are provided by external packages. We might as well come up with a contract at some point.

**Comment by janeklb** [comment]
> I've been working on our private monorepo setup with pnpm and have a lot of know-how now in this area. Not shareable in the current state, but I can contribute with thoughts, feedback and even some code from this experiment.
> 
> I think pnpm does things better in the monorepo / workspace tooling space than any other package manager.

@moltar If you have spare time and energy, please do share thoughts about your setup / projen + monorepos.  Are you using / have you taken any inspiration from `@aws/pdk` (which has a [monorepo](https://aws.github.io/aws-pdk/developer_guides/monorepo/index.html) construct)?

**Comment by moltar** [comment]
The MVP of a monorepo with pnpm is actually _very_ simple. All you need is a `pnpm-workspace.yaml` file with an array of `packages`. Typically a glob pattern is used there, but given that we control everything with projen, I am listing all of the sub-projects explicitly there. That is all you need to get a functional monorepo with `pnpm`. Pnpm will also provide multi-package command running (e.g. `pnpm run -r build`) and other niceties like `--filter` flag and `--concurrency`.

Another option you might want to set is `prefer-workspace-packages`, which will instruct pnpm to use packages from monorepo, rather than trying to install them from npm.

Another bonus feature related to pnpm is a full repo build command that correctly groups output (otherwise it is streamed line by line and gets quite messy, esp. in CI):

```sh
pnpm --if-present --recursive --aggregate-output --reporter=append-only build
```

Or Task expressed in `projen` like this:

```ts
exec: `${project.runScriptCommand} --if-present --recursive --aggregate-output --reporter=append-only ${project.buildTask.name}`,
```

^^ That is what I'd say the minimum is for pnpm to work.

---

Everything else is _not_ related to a package manager, but rather implementing/fixing Node.js and TypeScript ecosystem fragmentation and build tooling.

A really [good walkthru of what is required is by Moonrepo](https://moonrepo.dev/docs/guides/javascript/typescript-project-refs). In all of the years of working with TS, their guide is truly comprehensive and the only one I found that covers all angles of TS build system in the monorepo context.

---

Our proprietary solution currently includes the following components/tools that are all wired up to work in tandem:

- Changesets for npm publishing
- Husky
- Lint Staged
- Plop
- Pnpm
- Renovate Bot (self-hosted via GH actions)
- secret lint
- Syncpack (enforce consistency of dependency versions across monorepo)
- Better package building for TS, which includes:
	- `tsup` for bundling
	- Correct package.json exports, that supports CJS and ESM
	- Validation with `arethetypeswrong`
	- Validation with `publint`
	- Cleaning of `package.json` from cruft w/ `clean-publish`
- TS path mapping across all tools (TS, Jest, Vitest, tsup)
- TS path mapping (linking) monorepo projects, so that you don't need to build each change, can just `import`.
- TS project references
- Vitest (instead of Jest)
- VS Code Workspace (multi-folder setup).

---

I cannot share all of this as-is, unfortunately, but I can answer any questions you may have about any of the tools.

**Comment by skyrpex** [comment]
I have a set of projen projects for pnpm monorepos here https://github.com/skyrpex/wingen. There's no documentation but:

- Try it running `pnpx projen new --from @skyrpex/wingen monorepo-project --name @my/monorepo
`
- Allows you to create [NodeCjsProject](https://github.com/skyrpex/wingen/blob/main/src/node-cjs-project.ts), [NodeEsmProject](https://github.com/skyrpex/wingen/blob/main/src/node-esm-project.ts) and [TypeScriptProject](https://github.com/skyrpex/wingen/blob/main/src/typescript-project.ts)
- Very opinionated at the moment
- Uses turborepo, editorconfig, eslint, prettier, tsup, vitest and tsx
- Had to copy & modify certain parts of Projen's core to work around an issue regarding [pnpm's lowest matching policy back then](https://github.com/pnpm/pnpm/issues/6463)
